cpp/src/text/subword/wordpiece_tokenizer.cu

```diff
@@ -276,59 +278,40 @@ wordpiece_tokenizer::wordpiece_tokenizer(hashed_vocabulary const& vocab_table,
-  cub::DeviceSelect::If(nullptr,
-                        temp_storage_bytes,
-                        device_word_indices.data(),
-                        device_word_indices.data(),
-                        device_num_selected.data(),
-                        device_word_indices_count,
-                        select_op,
-                        stream);
-  cub::DeviceScan::InclusiveSum(nullptr,
-                                temp_storage_bytes_2,
-                                device_tokens_per_word.data(),
-                                device_word_indices.data(),
-                                max_new_char_total,
-                                stream);
-  max_cub_storage_bytes = std::max(temp_storage_bytes, temp_storage_bytes_2);
-  cub_temp_storage.resize(max_cub_storage_bytes, stream);
 }
 
-std::pair<uint32_t*, uint32_t*> wordpiece_tokenizer::tokenize(char const* d_strings,
-                                                              uint32_t const* d_offsets,
-                                                              uint32_t num_strings,
-                                                              cudaStream_t stream)
+uvector_pair wordpiece_tokenizer::tokenize(char const* d_strings,
+                                           uint32_t const* d_offsets,
+                                           uint32_t num_strings,
+                                           cudaStream_t stream)
 {
   auto cps_and_offsets = normalizer.normalize(d_strings, d_offsets, num_strings, stream);
-  tokenize(cps_and_offsets.first, cps_and_offsets.second, stream);
-  return std::make_pair(cps_and_offsets.first.gpu_ptr, cps_and_offsets.second.gpu_ptr);
+  tokenize(cps_and_offsets, stream);
+  return uvector_pair(std::move(cps_and_offsets.first), std::move(cps_and_offsets.second));
 }
 
-void wordpiece_tokenizer::tokenize(ptr_length_pair& cp_and_length,
-                                   ptr_length_pair& offsets_and_length,
-                                   cudaStream_t stream)
+struct copy_if_fn {  // inline lambda not allowed in private or protected member function
+  __device__ bool operator()(uint32_t cp) { return cp != std::numeric_limits<uint32_t>::max(); }
+};
+
+struct tranform_fn {  // just converting uint8 value to uint32
+  __device__ uint32_t operator()(uint8_t count) { return count; }
+};
+
+void wordpiece_tokenizer::tokenize(uvector_pair& cps_and_offsets, cudaStream_t stream)
 {
-  uint32_t* device_code_points     = cp_and_length.gpu_ptr;
-  size_t const num_code_points     = cp_and_length.length;
-  uint32_t* device_strings_offsets = offsets_and_length.gpu_ptr;
-  uint32_t const num_strings       = offsets_and_length.length - 1;
+  uint32_t* device_code_points     = cps_and_offsets.first->data();
+  size_t const num_code_points     = cps_and_offsets.first->size();
+  uint32_t* device_strings_offsets = cps_and_offsets.second->data();
+  uint32_t const num_strings       = cps_and_offsets.second->size() - 1;
```

Comment:
```suggestion
  uint32_t* const device_code_points     = cps_and_offsets.first->data();
  size_t const num_code_points           = cps_and_offsets.first->size();
  uint32_t* const device_strings_offsets = cps_and_offsets.second->data();
  uint32_t const num_strings             = cps_and_offsets.second->size() - 1;
```

Reply:
These 2 buffers are repurposed as working memory later in the function and so cannot be made const.

Reply:
From what I can see, only the buffer content is modified, the pointers are not reseated. So the pointers can be const and point to non-const data.

Reply:
Technically true 😊
Now I’m worried someone make the same mistake I did. I’d rather leave it as is for clarity if that is ok.

Reply:
I see what you mean. The syntax is very unintuitive here. I find the const helpful here as it tells me a lot - the variable will always point to the same memory, whose content might change.
I'm not really insisting on the change, we don't really apply const that systematically. I would not have left a comment here if not for the other two consts :)

Reply:
I don't think worrying that people will not know the rules is a good reason not to follow them. :)